commit 3cea1891748e0ed8e79fa5d9afe40750319751d1
Author: Aaron Lewis <aaronlewis@google.com>
Date:   Mon Oct 12 12:47:16 2020 -0700

    selftests: kvm: Test MSR exiting to userspace
    
    Add a selftest to test that when the ioctl KVM_X86_SET_MSR_FILTER is
    called with an MSR list, those MSRs exit to userspace.
    
    This test uses 3 MSRs to test this:
      1. MSR_IA32_XSS, an MSR the kernel knows about.
      2. MSR_IA32_FLUSH_CMD, an MSR the kernel does not know about.
      3. MSR_NON_EXISTENT, an MSR invented in this test for the purposes of
         passing a fake MSR from the guest to userspace.  KVM just acts as a
         pass through.
    
    Userspace is also able to inject a #GP.  This is demonstrated when
    MSR_IA32_XSS and MSR_IA32_FLUSH_CMD are misused in the test.  When this
    happens a #GP is initiated in userspace to be thrown in the guest which is
    handled gracefully by the exception handling framework introduced earlier
    in this series.
    
    Tests for the generic instruction emulator were also added.  For this to
    work the module parameter kvm.force_emulation_prefix=1 has to be enabled.
    If it isn't enabled the tests will be skipped.
    
    A test was also added to ensure the MSR permission bitmap is being set
    correctly by executing reads and writes of MSR_FS_BASE and MSR_GS_BASE
    in the guest while alternating which MSR userspace should intercept.  If
    the permission bitmap is being set correctly only one of the MSRs should
    be coming through at a time, and the guest should be able to read and
    write the other one directly.
    
    Signed-off-by: Aaron Lewis <aaronlewis@google.com>
    Reviewed-by: Alexander Graf <graf@amazon.com>
    Message-Id: <20201012194716.3950330-5-aaronlewis@google.com>
    Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

diff --git a/tools/testing/selftests/kvm/lib/kvm_util.c b/tools/testing/selftests/kvm/lib/kvm_util.c
index b2c426adb87f..88ef7067f1e6 100644
--- a/tools/testing/selftests/kvm/lib/kvm_util.c
+++ b/tools/testing/selftests/kvm/lib/kvm_util.c
@@ -1806,6 +1806,8 @@ static struct exit_reason {
 	{KVM_EXIT_OSI, "OSI"},
 	{KVM_EXIT_PAPR_HCALL, "PAPR_HCALL"},
 	{KVM_EXIT_DIRTY_RING_FULL, "DIRTY_RING_FULL"},
+	{KVM_EXIT_X86_RDMSR, "RDMSR"},
+	{KVM_EXIT_X86_WRMSR, "WRMSR"},
 #ifdef KVM_EXIT_MEMORY_NOT_PRESENT
 	{KVM_EXIT_MEMORY_NOT_PRESENT, "MEMORY_NOT_PRESENT"},
 #endif